text/0000-router-helpers.md

```diff
@@ -0,0 +1,571 @@
+- **`queryParms`** _Object_: Optionally pass key value pairs that will be used to determine if the route is loading.
+
+_Returns_
+- _Boolean_: Determines if the route is loading or not.
+
+### `{{is-transitioning-in}}` Helper
+
+```hbs
+{{is-transitioning-in 'people.index' model queryParams=(hash a=a)}}
+```
+
+Or
+
+```hbs
+{{is-transitioning-in 'people.index' models=model queryParams=(hash a=a)}}
+```
+
+The arguments to `{{is-transitioning-in}}` have the same semantics as all the other route state helpers, however `{{is-transitioning-in}}` only returns `true` when the route is going from an non-active to an active state. This should provide the same logic that determines whether to put an `ember-transition-in` class on a `{{link-to}}`.
+
+#### Signature Explainer
+
+Using the example above:
+
+_Positional Params_
+- **`routeName`** _String_: A fully-qualified name of this route, like `"people.index"`
+- **`model`** _...Object|Array|Number|String_: Optionally pass an arbitrary amount of models to use for generation
+
+_Named Params_
+- **`models`** _Object|Array|Number|String_: Same as the positional parameter. It is a compiler error if the positional params and the named param is used.
+- **`queryParms`** _Object_:  Optionally pass key value pairs that will be used to determine if the route is transitioning in.
+
+_Returns_
+- _Boolean_: Determines if the route is transitioning in.
+
+### `{{is-transitioning-out}}` Helper
+
+```hbs
+{{is-transitioning-out 'people.index' model queryParams=(hash a=a)}}
+```
+
+Or
+
+```hbs
+{{is-transitioning-out 'people.index' models=model queryParams=(hash a=a)}}
+```
+
+`{{is-transitioning-out}}` is just the inverse of `{{is-transitioning-in}}`.
+
+#### Signature Explainer
+
+Using the example above:
+
+_Positional Params_
+- **`routeName`** _String_: A fully-qualified name of this route, like `"people.index"`
+- **`model`** _...Object|Array|Number|String_: Optionally pass an arbitrary amount of models to use for generation
+
+_Named Params_
+- **`models`** _Object|Array|Number|String_: Same as the positional parameter. It is a compiler error if the positional params and the named param is used.
+- **`queryParms`** _Object_:  Optionally pass key value pairs that will be used to determine if the route is transitioning out.
+
+_Returns_
+- _Boolean_: Determines if the route is transitioning out.
+
+### Event Dispatcher
+
+In the past, only `HTMLAnchorElement`s that were produced by `{{link-to}}`s would produce a transition when a user clicked on them. This RFC changes to the global `EventDispatcher` to allow for any `HTMLAnchorElement` with a valid root relative `href` to cause a transition. This will allow for us to not only allows us to support use cases like the ones described in the [motivation](#anchor-tags), it makes teaching easier since people who know HTML don't need know an Ember specific API to participate in routing transitions.
```

Comment:
@chadhietala A question for you here. Everything else on this RFC looks awesome and I'm quite eager to see it all land. When it comes to this change though, I'd like to raise a potential snag we should make sure we address.

Specifically, in a "switching into Ember" scenario for a primarily server-rendered app, will there be enough safety valves as part of this change to allow folks to safely drop Ember in on their site without it unexpectedly hijacking all `<a>` tags? If there weren't a wildcard route setup as specified below, theoretically the Ember router would still be in a `whitelist` scenario and would only intercept click events for `<a>` tags that it **has** been setup to handle. But thought it was worth clarifying so we don't inadvertently cause ourselves problems for our `npm install your way to Ember` story down the road ...

Reply:
The event dispatcher is only hooked up on the ember application `rootElement`, so it's fine to drop in Ember on a server-rendered page, even after this change.

https://github.com/emberjs/ember.js/blob/aa838632/packages/%40ember/-internals/views/lib/system/event_dispatcher.js#L241

Reply:
@acorncom Yes they should continue to work in an SSR'd page since the events are rooted on the `rootElement`. The app that is migrating needs to make sure the root element is not `body`. I think this is typically what people to do in this case.